src/main/java/com/arpnetworking/clusteraggregator/client/AggClientConnection.java

```diff
@@ -191,9 +193,57 @@ private void processMessages() {
         _buffer = current;
     }
 
-    private PeriodicData buildPeriodicData(final Messages.StatisticSetRecord setRecord) {
+    private Optional<PeriodicData> buildPeriodicData(final Messages.StatisticSetRecord setRecord) {
         final CombinedMetricData combinedMetricData = CombinedMetricData.Builder.fromStatisticSetRecord(setRecord).build();
         final ImmutableList.Builder<AggregatedData> builder = ImmutableList.builder();
+        final ImmutableMap.Builder<String, String> dimensionBuilder = ImmutableMap.builder();
+
+        Optional<String> host = Optional.absent();
+        Optional<String> service = Optional.absent();
+        Optional<String> cluster = Optional.absent();
+        for (Messages.DimensionEntry dimensionEntry : setRecord.getDimensionsList()) {
+            if (HOST_KEY.equals(dimensionEntry.getKey())) {
+                host = Optional.fromNullable(dimensionEntry.getValue());
+            } else if (SERVICE_KEY.equals(dimensionEntry.getKey())) {
+                service = Optional.fromNullable(dimensionEntry.getValue());
+            } else if (CLUSTER_KEY.equals(dimensionEntry.getKey())) {
+                cluster = Optional.fromNullable(dimensionEntry.getValue());
+            } else {
+                dimensionBuilder.put(dimensionEntry.getKey(), dimensionEntry.getValue());
+            }
+        }
+
+        if (!service.isPresent()) {
+            service = Optional.fromNullable(setRecord.getService());
+        }
+
+        if (!cluster.isPresent()) {
+            cluster = Optional.fromNullable(setRecord.getCluster());
+            if (!cluster.isPresent()) {
+                cluster = _clusterName;
+            }
+        }
+
+        if (!host.isPresent()) {
+            host = _hostName;
+        }
+
+        dimensionBuilder.put(HOST_KEY, host.or(""));
+        dimensionBuilder.put(SERVICE_KEY, service.or(""));
+        dimensionBuilder.put(CLUSTER_KEY, cluster.or(""));
+
+        if (!(host.isPresent() && service.isPresent() && cluster.isPresent())) {
+            INCOMPLETE_RECORD_LOGGER.warn()
```

Comment:
Normally, I would prefer this be logged in the caller since that is where the "drop" part is actually realized (e.g. based on absent return value). However, we don't know the cause of the absent in the caller. Any thoughts?